src/main/java/bio/terra/service/filedata/azure/util/BlobContainerCopierBuilder.java

```diff
@@ -0,0 +1,145 @@
+package bio.terra.service.filedata.azure.util;
+
+import com.azure.storage.blob.BlobUrlParts;
+import java.time.Duration;
+import java.util.List;
+import java.util.Objects;
+import org.apache.commons.lang.StringUtils;
+import org.apache.logging.log4j.util.Strings;
+
+/** Builder for the {@link BlobContainerCopier} */
+public final class BlobContainerCopierBuilder {
+
+  public static final int DEFAULT_LIST_OPERATION_TIMEOUT_IN_SECONDS = 30;
+  public static final int DEFAULT_POLLING_INTERVAL_IN_SECONDS = 5;
+  private static final int MIN_POLLING_INTERVAL = 2;
+  private static final int MIN_LIST_TIMEOUT = 10;
+
+  private BlobContainerClientFactory sourceClientFactory;
+  private BlobContainerClientFactory destinationClientFactory;
+  private String sourceContainerPrefix;
+  private String sourceBlobUrl;
+  private Duration pollingInterval = Duration.ofSeconds(DEFAULT_POLLING_INTERVAL_IN_SECONDS);
+  private Duration listOperationTimeout =
+      Duration.ofSeconds(DEFAULT_LIST_OPERATION_TIMEOUT_IN_SECONDS);
+  private List<BlobCopySourceDestinationPair> sourceDestinationPairs;
+  private String destinationBlobName;
+
+  public BlobContainerCopierBuilder() {
+    this.sourceContainerPrefix = "";
+  }
+
+  public BlobContainerCopierBuilder destinationBlobName(String destinationBlobName) {
+    this.destinationBlobName = destinationBlobName;
+    return this;
+  }
+
+  public BlobContainerCopierBuilder sourceDestinationPairs(
+      List<BlobCopySourceDestinationPair> pairs) {
+    this.sourceDestinationPairs = pairs;
+    return this;
+  }
+
+  public BlobContainerCopierBuilder sourceClientFactory(BlobContainerClientFactory factory) {
+    this.sourceClientFactory =
```

Comment:
This is the same pattern that the SDK follows in their builders. The idea is that you want to validate them upon getting them set for optional properties, which indicates that the caller wants to use them, and therefore the caller must provide valid values. If you wait until the build time, it is hard to determine if a null/empty value exists because the caller didn't intend to use the property or a bug/unexpected behavior incorrectly set such the property.

Reply:
I guess the idea is that you don't do the validation until you've got a complete set of parameters, which seems better to me